feat: v0.3.0 Disappearing Messages support

yougotwill opened the pull request:
We can't add new arguments to the getUserInfo and setUserInfo functions because it would break migrations

Edit: I've realised it's probably best to only merge this once I'm done integrating disappearing messages so will keep updating this branch.

yougotwill committed:
feat: updated the user profile wrapper to use setExpiry and getExpiry

We can't add new arguments to the getUserInfo and setUserInfo functions because it would break migrations

jagerman commented on src/user_config.hpp:
```diff
@@ -22,6 +22,9 @@ class UserConfigWrapper : public ConfigBaseImpl, public Napi::ObjectWrap<UserCon
     void setEnableBlindedMsgRequest(const Napi::CallbackInfo& info);
+
+    Napi::Value getExpiry(const Napi::CallbackInfo& info);
+    void setExpiry(const Napi::CallbackInfo& info);
```
These are a bit confusingly named, since this is specifically setting/returning the note-to-self expiry rather than a more general User Config expiry that one might guess from the name.

yougotwill replied:
My understanding is still beginner level so forgive any incorrect assumptions I make. But I had a look through the code and couldn't spot what you meant by "general User Config expiry" the only expiry methods in User Profile are the above functions. Which are exclusive to Note to Self since that is what the User Profile is? The rest is handled in Contacts. 

Or are you talking about making the function name more explicit like "setNoteToSelfExpiry"? Audric and I chatted about this yesterday and are trying to keep the wrapper method signatures as close to the original libsession functions as possible. We will only call these functions in relation to the UserProfile so we know it is Note to Self through that.

Example session-desktop call: `userProfileWrapper.setExpiry(expirySeconds);`

jagerman commented:
> Or are you talking about making the function name more explicit like "setNoteToSelfExpiry"? Audric and I chatted about this yesterday and are trying to keep the wrapper method signatures as close to the original libsession functions as possible

Yeah, this.  The C++ API function is `set_nts_expiry`.

jagerman commented:
Where `nts` = `note to self` but without being Java.

yougotwill commented:
@jagerman That makes sense! I didn't pick up on the `nts` will update to make the names more explicit. Thanks!

yougotwill committed:
fix: made expiry functions in user config more explicit to NoteToSelf

yougotwill committed:
feat: added expirationMode and expirationTimerSeconds to contacts set function

yougotwill committed:
fix: export DisappearingMessageConversationType

yougotwill committed:
fix: change disappearAfter to deleteAfter to match session desktop

yougotwill committed:
feat: disappearing messages in legacy groups

yougotwill committed:
fix: updated setLegacyGroup function

yougotwill committed:
fix: check if setLegacyGroup arg is empty and throw

yougotwill committed:
feat: comment for getNoteToSelfExpiry

yougotwill committed:
Merge branch 'main' into feat/ses-60/set_and_get_expiry

yougotwill committed:
refactor: DisappearingMessageConversationType to DisappearingMessageConversationModeType

this matches session-desktop

yougotwill committed:
feat: comments and made a note in usergroups

yougotwill committed:
chore: bump to version 0.3.0

since we have changed function signatures for disappearing messages we do a minor version bump. README explains the expected behaviour

yougotwill commented:
This PR includes all code changes required for Disappearing Messages V2 to work on Session Desktop. Since it changes some existing function signatures we have done a minor version bump

yougotwill committed:
fix: duplicated deleteAfterRead instead of deleteAfterSend

Bilb commented on src/user_config.cpp:
```diff
@@ -114,4 +105,25 @@ void UserConfigWrapper::setEnableBlindedMsgRequest(const Napi::CallbackInfo& inf
 }
 
+Napi::Value UserConfigWrapper::getNoteToSelfExpiry(const Napi::CallbackInfo& info) {
+    return wrapResult(info, [&] {
+        auto env = info.Env();
+        auto expirySeconds = toJs(env, config.get_nts_expiry()->count());
+
+        return expirySeconds;
+    });
+}
```
you don't need the toJs here, the wrapResuilts does it for you. 
you can just return `config.get_nts_expiry()->count()`

yougotwill replied:
Nice, didn't read the wrapResult properly. Fixed

Bilb commented on src/user_groups_config.cpp:
```diff
@@ -154,11 +154,14 @@ Napi::Value UserGroupsWrapper::getAllLegacyGroups(const Napi::CallbackInfo& info
+
         auto legacyGroupValue = info[0];
         assertIsObject(legacyGroupValue);
-
         auto obj = legacyGroupValue.As<Napi::Object>();
 
+        if (obj.IsEmpty())
+            throw std::invalid_argument("cppLegacyGroup received empty");
```
this obj.IsEmpty() check is already part of assertIsObject()

yougotwill replied:
Fixed 👍

yougotwill committed:
fix: getNoteToSelfExpiry doesnt need toJS

wrapResult converts the callable info for us

yougotwill committed:
fix: setLegacyGroup dont need is empty check

since we asset the object before

yougotwill committed:
feature: DisappearingMessageConversationModeType

added legacy type for now, will remove later

yougotwill committed:
feat: updated libsession to 1.1.0

merge function now returns good hashes
